nestjs/src/courses/team-distribution/team.controller.ts

```diff
@@ -0,0 +1,169 @@
+import { TeamDistributionService } from './team-distribution.service';
+import { TeamService } from './team.service';
+
+@Controller('courses/:courseId/team-distribution/:distributionId/team')
+@ApiTags('team')
+@UseGuards(DefaultGuard, CourseGuard)
+export class TeamController {
+  constructor(
+    private readonly teamService: TeamService,
+    private readonly studentService: StudentsService,
+    private readonly distributionService: TeamDistributionService,
+  ) {}
+
+  @Get('/')
+  @UseGuards(RoleGuard)
+  @ApiOkResponse({ type: TeamsDto })
+  @ApiOperation({ operationId: 'getTeams' })
+  @RequiredRoles([CourseRole.Student, Role.Admin, CourseRole.Manager])
+  public async getTeams(
+    @Param('courseId', ParseIntPipe) _: number,
+    @Param('distributionId', ParseIntPipe) distributionId: number,
+    @Query() query: TeamsQueryDto,
+  ) {
+    const page = parseInt(query.current);
+    const limit = parseInt(query.pageSize);
+    const { teams, paginationMeta } = await this.teamService.findByDistributionId(distributionId, {
+      page,
+      limit,
+    });
+
+    return new TeamsDto(teams, paginationMeta);
+  }
+
+  @Post('/')
+  @UseGuards(RoleGuard)
+  @ApiOkResponse({ type: TeamDto })
+  @ApiOperation({ operationId: 'createTeam' })
+  @RequiredRoles([CourseRole.Student, Role.Admin, CourseRole.Manager])
+  public async create(
+    @Req() req: CurrentRequest,
+    @Param('courseId', ParseIntPipe) courseId: number,
+    @Param('distributionId', ParseIntPipe) distributionId: number,
+    @Body() dto: CreateTeamDto,
+  ) {
+    const studentId = req.user.courses[courseId]?.studentId;
+    const students: Student[] = [];
+    const isManager = req.user.isAdmin || req.user.courses[courseId]?.roles.includes(CourseRole.Manager);
+
+    if (studentId && !isManager) {
+      const student = await this.studentService.getStudentDetailed(studentId);
+      if (student.teams.find(t => t.teamDistributionId === distributionId)) {
+        throw new BadRequestException();
+      }
+      students.push(student);
+    }
+    const data = await this.teamService.create({
+      teamDistributionId: distributionId,
+      students,
+      ...dto,
+    });
+    const team = await this.teamService.findByIdDetailed(data.id);
+
+    if (team.students.length) {
+      const distribution = await this.distributionService.getById(distributionId);
+      await Promise.all(
+        team.students.map(el => this.studentService.deleteStudentFromTeamDistribution(el.id, distribution)),
+      );
+    }
+
+    return new TeamDto(team);
+  }
+
+  @Patch('/:id')
+  @RequiredRoles([Role.Admin, CourseRole.Manager, CourseRole.Student])
+  @ApiOperation({ operationId: 'updateTeam' })
+  @ApiOkResponse()
+  public async updateTeam(
+    @Param('courseId', ParseIntPipe) _: number,
+    @Param('distributionId', ParseIntPipe) _distributionId: number,
+    @Param('id', ParseIntPipe) id: number,
+    @Body() dto: UpdateTeamDto,
+  ) {
+    await this.teamService.update(id, dto);
```

Comment:
I think it should have some validation on who is allowed to update team - not all students

Reply:
fixed